File: sdk/node/_binary.js
// chromix (Node) — binary management: detect platform, download the matching
// bundle from the GitHub Release, verify SHA256SUMS, cache under ~/.cache/chromix.
import { spawnSync } from "node:child_process";
import { createWriteStream, existsSync, chmodSync, mkdirSync, createReadStream } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";
import { pipeline } from "node:stream/promises";
import { createHash } from "node:crypto";

export const VERSION = "151.0.7922.174";
const REPO = "xiaozhou26/Chromix";
// Two release channels. "stable" = Chromium 149 (matches the version the mass of
// real users run). "latest" = 151 (newest engine). See build/versions.txt.
export const CHANNELS = {
  stable: { tag: "v149.0.7827.200" },
  latest: { tag: "v151.0.7922.174" },
};
export const CACHE = process.env.CHROMIX_CACHE_DIR || join(homedir(), ".cache", "chromix");
export const hostFor = (tag) => process.env.CHROMIX_DOWNLOAD_HOST
  || `https://github.com/${REPO}/releases/download/${tag}`;

// platform key -> { asset, kind, launcher }
export const ASSETS = {
  "linux-x64": { asset: "chromix-linux-x64.tar.gz", kind: "tar", launcher: "chromix/chromix" },
  "win-x64":   { asset: "chromix-win-x64.zip",      kind: "zip", launcher: "chromix/chromix.cmd" },
  "mac-arm64": { asset: "chromix-mac-arm64.tar.gz", kind: "tar", launcher: "chromix/chromix" },
  "mac-x64":   { asset: "chromix-mac-x64.tar.gz",   kind: "tar", launcher: "chromix/chromix" },
};

export function resolvePlatform() {
  const { platform, arch } = process;
  if (platform === "linux" && arch === "x64") return "linux-x64";
  if (platform === "win32" && arch === "x64") return "win-x64";
  if (platform === "darwin") return arch === "arm64" ? "mac-arm64" : "mac-x64";
  return null;
}

export async function sha256(path) {
  const h = createHash("sha256");
  await pipeline(createReadStream(path), h);
  return h.digest("hex");
}

export async function expectedSha(asset, host) {
  try {
    const r = await fetch(`${host}/SHA256SUMS`);
    if (!r.ok) return null;
    for (const line of (await r.text()).split("\n")) {
      const p = line.trim().split(/\s+/);
      if (p.length === 2 && p[1].replace(/^\*/, "") === asset) return p[0].toLowerCase();
    }
  } catch { /* none */ }
  return null;
}

export async function ensureNative(plat, host, tag) {
  const { asset, kind, launcher } = ASSETS[plat];
  const root = join(CACHE, tag, plat);   // cache per release tag so channels don't collide
  const launcherPath = join(root, launcher);
  if (existsSync(launcherPath)) return launcherPath;
  mkdirSync(root, { recursive: true });
  const archive = join(root, asset);
  process.stderr.write(`[chromix] downloading ${host}/${asset} ...\n`);
  const res = await fetch(`${host}/${asset}`);
  if (!res.ok) throw new Error(`download failed: ${res.status}`);
  await pipeline(res.body, createWriteStream(archive));

  const exp = await expectedSha(asset, host);
  if (exp) {
    const act = await sha256(archive);
    if (act !== exp) throw new Error(`SHA256 mismatch for ${asset}: expected ${exp}, got ${act}`);
    process.stderr.write("[chromix] SHA256 verified\n");
  } else {
    process.stderr.write("[chromix] WARNING: no SHA256SUMS published; skipping verification\n");
  }

  if (kind === "tar") {
    if (spawnSync("tar", ["xzf", archive, "-C", root], { stdio: "inherit" }).status !== 0)
      throw new Error("tar extraction failed");
  } else { // zip (Windows): use PowerShell Expand-Archive
    if (spawnSync("powershell", ["-NoProfile", "-Command",
        `Expand-Archive -Force -LiteralPath '${archive}' -DestinationPath '${root}'`],
        { stdio: "inherit" }).status !== 0) throw new Error("zip extraction failed");
  }
  if (!launcher.endsWith(".cmd") && existsSync(launcherPath)) chmodSync(launcherPath, 0o755);
  if (!existsSync(launcherPath)) throw new Error(`launcher missing after extract: ${launcherPath}`);
  return launcherPath;
}

File: sdk/python/chromix/_binary.py
"""Binary management for chromix: download, verify, cache the stealth Chromium bundle.

Detects the platform, downloads the matching bundle from the GitHub Release,
verifies it against SHA256SUMS, and caches it under ``~/.cache/chromix``.
"""
from __future__ import annotations
import hashlib
import os
import platform
import sys
import tarfile
import urllib.request
import zipfile
from pathlib import Path

_REPO = "xiaozhou26/Chromix"
# Two release channels. "stable" = Chromium 149 (matches the version the mass of
# real users run). "latest" = 152 (newest engine). See build/versions.txt.
_CHANNELS = {
    "stable": {"tag": "v149.0.7827.200"},
    "latest": {"tag": "v152.0.7977.75"},
}
_CACHE = Path(os.environ.get("CHROMIX_CACHE_DIR",
                             Path.home() / ".cache" / "chromix"))


def _host(tag: str) -> str:
    return os.environ.get("CHROMIX_DOWNLOAD_HOST",
                          f"https://github.com/{_REPO}/releases/download/{tag}")


# platform key -> (release asset, archive kind, launcher relative path)
_ASSETS = {
    "linux-x64":  ("chromix-linux-x64.tar.gz", "tar", "chromix/chromix"),
    "win-x64":    ("chromix-win-x64.zip",      "zip", "chromix/chromix.cmd"),
    "mac-arm64":  ("chromix-mac-arm64.tar.gz", "tar", "chromix/chromix"),
    "mac-x64":    ("chromix-mac-x64.tar.gz",   "tar", "chromix/chromix"),
}


def resolve_platform() -> str | None:
    sysname, mach = platform.system(), platform.machine().lower()
    if sysname == "Linux" and mach in ("x86_64", "amd64"):
        return "linux-x64"
    if sysname == "Windows" and mach in ("amd64", "x86_64"):
        return "win-x64"
    if sysname == "Darwin":
        return "mac-arm64" if mach in ("arm64", "aarch64") else "mac-x64"
    return None


def _sha256(path: Path) -> str:
    h = hashlib.sha256()
    with open(path, "rb") as fh:
        for chunk in iter(lambda: fh.read(1 << 20), b""):
            h.update(chunk)
    return h.hexdigest()


def _expected_sha(asset: str, host: str) -> str | None:
    """Fetch SHA256SUMS from the release and return the hash for `asset`."""
    try:
        with urllib.request.urlopen(f"{host}/SHA256SUMS", timeout=30) as r:
            for line in r.read().decode().splitlines():
                parts = line.split()
                if len(parts) == 2 and parts[1].lstrip("*") == asset:
                    return parts[0].lower()
    except Exception:
        return None
    return None


def _download(plat: str, host: str, tag: str) -> Path:
    """Ensure the bundle for `plat` is present + verified; return the launcher path."""
    asset, kind, launcher_rel = _ASSETS[plat]
    root = _CACHE / tag / plat   # cache per release tag so channels don't collide
    launcher = root / launcher_rel
    if launcher.exists():
        return launcher
    root.mkdir(parents=True, exist_ok=True)
    archive = root / asset
    url = f"{host}/{asset}"
    sys.stderr.write(f"[chromix] downloading {url} ...\n")
    urllib.request.urlretrieve(url, archive)

    expected = _expected_sha(asset, host)
    if expected:
        actual = _sha256(archive)
        if actual != expected:
            archive.unlink(missing_ok=True)
            raise RuntimeError(f"SHA256 mismatch for {asset}: expected {expected}, got {actual}")
        sys.stderr.write("[chromix] SHA256 verified\n")
    else:
        sys.stderr.write("[chromix] WARNING: no SHA256SUMS published; skipping verification\n")

    if kind == "tar":
        with tarfile.open(archive) as t:
            t.extractall(root)
    else:
        with zipfile.ZipFile(archive) as z:
            z.extractall(root)
    archive.unlink(missing_ok=True)
    if launcher.exists() and not launcher.name.endswith(".cmd"):
        launcher.chmod(0o755)
    if not launcher.exists():
        raise RuntimeError(f"bundle extracted but launcher missing: {launcher}")
    return launcher
